File: alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/models/DAOModelRegistrar.java
package com.github.dynamicextensionsalfresco.models;

import java.util.Iterator;
import java.util.List;
import java.util.LinkedList;
import org.alfresco.repo.dictionary.DictionaryDAO;
import org.alfresco.service.namespace.QName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Registers and unregisters [M2Model]s with a [DictionaryDAO]. This class is stateful and intended to
 * support dynamic loading of models.
 *
 * @author Laurens Fridael
 * @author Toon Geens
 */

public class DAOModelRegistrar extends AbstractModelRegistrar {

    private final Logger logger = LoggerFactory.getLogger(DAOModelRegistrar.class);

    private final DictionaryDAO dictionaryDao;

    /* State */
    private final List<QName> registeredModelNames = new LinkedList<>();

    public DAOModelRegistrar(DictionaryDAO dictionaryDao, M2ModelListProvider modelsToRegister) {
        super(modelsToRegister);
        this.dictionaryDao = dictionaryDao;
    }

    @Override
    protected void registerModel(M2ModelResource modelResource) {
        QName qname = dictionaryDao.putModel(modelResource.getM2Model());
        registeredModelNames.add(qname);
    }

    @Override
    public void unregisterModels() {
        Iterator<QName> it = this.registeredModelNames.iterator();
        while (it.hasNext()) {
            QName modelName = it.next();
            try {
                logger.debug("Unregistering model '{}' via DictionaryDAO", modelName);
                dictionaryDao.removeModel(modelName);
            } finally {
                it.remove();
            }
        }
    }
}

File: alfresco-integration/src/main/kotlin/com/github/dynamicextensionsalfresco/models/DAOModelRegistrar.kt
package com.github.dynamicextensionsalfresco.models

import org.alfresco.repo.dictionary.DictionaryDAO
import org.alfresco.repo.dictionary.M2Model
import org.alfresco.service.namespace.QName
import org.slf4j.LoggerFactory
import java.util.*

/**
 * Registers and unregisters [M2Model]s with a [DictionaryDAO]. This class is stateful and intended to
 * support dynamic loading of models.

 * @author Laurens Fridael
 */
public open class DAOModelRegistrar(private val dictionaryDao: DictionaryDAO, modelsToRegister: M2ModelListProvider) : AbstractModelRegistrar(modelsToRegister) {

    private val logger = LoggerFactory.getLogger(javaClass)

    /* State */

    private val registeredModelNames = LinkedList<QName>()

    /* Main operations */

    override fun unregisterModels() {
        val it = registeredModelNames.iterator()
        while (it.hasNext()) {
            val modelName = it.next()
            try {
                logger.debug("Unregistering model '{}' via DictionaryDAO", modelName)
                dictionaryDao.removeModel(modelName)
            } finally {
                it.remove()
            }
        }
    }

    override fun registerModel(modelResource: M2ModelResource) {
        val qName = dictionaryDao.putModel(modelResource.m2Model)
        registeredModelNames.add(qName)
    }
}
